pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -418,7 +418,11 @@ public synchronized PinotResourceManagerResponse addInstance(Instance instance)
     if (instances.contains(instanceIdToAdd)) {
       return PinotResourceManagerResponse.failure("Instance " + instanceIdToAdd + " already exists");
     } else {
-      _helixAdmin.addInstance(_helixClusterName, InstanceUtils.toHelixInstanceConfig(instance));
+      InstanceConfig instanceConfig = InstanceUtils.toHelixInstanceConfig(instance);
+      if (instance.isQueriesDisabled()) {
```

Comment:
Can you move this part into `InstanceUtils.toHelixInstanceConfig()` so that all the conversion happens in a single place?